src/mococrw/mac.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#pragma once
+#include <memory>
+#include <vector>
+#include "openssl_wrap.h"
+
+namespace mococrw
+{
+
+/**
+ * @brief The MessageAuthenticationCode class is the abstract base class for the different implementation of
+ * message authentication codes.
+ */
+class MessageAuthenticationCode {
+public:
+    /**
+     * Adds the message to the hash. This function may be invoked multiple times.
+     * For finishing the calculation of the MAC-tag invoke finish().
+     *
+     * @param message chunk of data used for MAC.
+     * @throws MoCOCrWException if this function is invoked after finish was called.
+     */
+    void update(const std::vector<uint8_t>& message);
+
+    /**
+     * @brief Finalize the HMAC.
+     *
+     * This function calculates the hashed messaged authentication code.
+     *
+     * @throws MoCOCrWException if this function is invoked twice.
+     * @return The hashed message authentication code tag (MAC-tag)
+     */
+    std::vector<uint8_t> finish();
+};
+
+class HMAC : public MessageAuthenticationCode{
+public:
+    /**
+     * @brief Constructor
+     * @param hashFunction The hash function which shall be used
+     * @param tagBytesLength The length of the tag/mac in number of bytes (multiple of 8 bits)
+     */
+    HMAC(mococrw::openssl::DigestTypes hashFunction, const std::vector<uint8_t> &key, size_t outputLength);
+
+    ~HMAC();
+
+    /**
+     * Adds the chunk of data to the hash. H(i-key || message1 [ || message2])
+     * For getting the HMAC-tag invoke finish()
+     *
+     * @param message chunk of data to used for HMAC.
+     * @throws MoCOCrWException if this function is invoked after finish was called.
+     */
+    void update(const std::vector<uint8_t>& message);
+
+
+    /**
+     * @brief Finalize the HMAC.
+     *
+     * This function calculates the hashed messaged authentication code tag (HMAC-tag).
+     * H(o-key || H(i-key || message [ || message]))
+     *
+     * @throws MoCOCrWException if this function is invoked twice.
+     * @return The hashed message authentication code tag (HMAC-tag)
+     */
+    std::vector<uint8_t> finish();
+
+    /**
+     * @brief The copy constructor
+     * @param other
+     */
+    HMAC(const HMAC &other);
```

Comment:
Do we need (and/or want) this to be copyable?

Reply:
If there is no copy operator something like 
auto hmac = mococrw::HMAC(hashFunction, utility::fromHex(testData.key), testData.outputLength);

Fails with:
error: use of deleted function ‘mococrw::HMAC::HMAC(const mococrw::HMAC&)

MoCOCrW/tests/unit/test_mac.cpp:44: error: call to implicitly-deleted copy constructor of 'mococrw::HMAC'

Reply:
But
`mococrw::HMAC` hmac{hashFunction, utility::fromHex(testData.key};`
works.
If you want to support almost-always-auto a move constructor should suffice. All copy operations, I would honestly delete on these kind of objects.

Reply:
Removed the copy constructors and added move constructors and assignment operators.